File: shipman.Server/Domain/Rules/ShipmentEventFlow.cs
using shipman.Server.Domain.Enums;

namespace shipman.Server.Domain.Rules;

public static class ShipmentEventFlow
{
    public static readonly Dictionary<ShipmentEventType, ShipmentEventType[]> AllowedTransitions =
        new()
        {
            { ShipmentEventType.Created, new[] { ShipmentEventType.Prepared, ShipmentEventType.Cancelled } },
            { ShipmentEventType.Prepared, new[] { ShipmentEventType.HandedOver, ShipmentEventType.Cancelled, ShipmentEventType.Delayed } },
            { ShipmentEventType.HandedOver, new[] { ShipmentEventType.Delivered, ShipmentEventType.Delayed } },
            { ShipmentEventType.Delivered, Array.Empty<ShipmentEventType>() },
            { ShipmentEventType.Delayed, new[] { ShipmentEventType.Prepared, ShipmentEventType.HandedOver, ShipmentEventType.Delivered, ShipmentEventType.Cancelled } },
            { ShipmentEventType.Cancelled, Array.Empty<ShipmentEventType>() }
        };
}

File: shipman.client/src/features/shipments/shipmentEventFlow.ts
/**
 * Allowed next event types from current shipment status (mirrors ShipmentEventFlow on the server).
 */
const ALLOWED_NEXT: Record<string, string[]> = {
    Created: ['Prepared', 'Cancelled'],
    Prepared: ['HandedOver', 'Cancelled', 'Delayed'],
    HandedOver: ['Delivered', 'Delayed'],
    Delivered: [],
    Delayed: ['Prepared', 'HandedOver', 'Delivered', 'Cancelled'],
    Cancelled: []
};

export function getAllowedShipmentEventValues(currentStatus: string): string[]
{
    return ALLOWED_NEXT[currentStatus] ?? [];
}
